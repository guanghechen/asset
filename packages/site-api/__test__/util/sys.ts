/**
 * Create a promise resolved after ${milliseconds} ms.
 *
 * @param milliseconds
 */
export function noop(milliseconds: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, milliseconds)
  })
}
