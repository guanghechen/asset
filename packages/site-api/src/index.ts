export * from './config/deploy'
export * from './config/site'
export * from './config/sub-site/config'
export * from './config/sub-site/source-item'
export * from './core/entity/_types'
export * from './core/entity/asset'
export * from './core/entity/asset'
export * from './core/entity/category'
export * from './core/entity/tag'
export * from './core/manager/asset'
export * from './core/manager/asset-entity'
export * from './core/manager/category'
export * from './core/manager/tag'
export * from './core/parser'
export * from './core/processor'
export * from './core/provider'
export * from './core/service/asset'
export * from './core/service/category'
export * from './core/service/tag'
export * from './data/site'
export * from './util/fs'
export * from './util/path'
export * from './util/string'
export * from './util/sync'
export * from './util/uuid'
