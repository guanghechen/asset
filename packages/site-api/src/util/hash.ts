import crypto from 'crypto'

/**
 * Calc hash value with sha1 algorithm
 *
 * @param content
 */
export function sha1(content: string | Buffer): string {
  const sha1 = crypto.createHash('sha1')
  sha1.update(content)
  return sha1.digest('hex')
}


/**
 * calc fingerprint of content
 * @param content
 */
export function calcFingerprint(content: Buffer): string {
  return sha1(content)
}


/**
 * Generate unique id from text
 *
 * @param text
 */
export function uniqueText(
  text: string | null,
): typeof text extends null ? null : string {
  if (text == null) return null as any
  return text.toLowerCase().trim().replace(/[\s\n]+/g, '--')
}
