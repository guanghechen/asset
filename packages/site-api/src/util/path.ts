import path from 'path'


/**
 * Resolve url path
 *
 * @param prefixPath
 * @param p
 */
export function resolveUrlPath(
  prefixPath: string,
  ...pathPieces: string[]
): string {
  const result = (prefixPath + '/' + pathPieces.join('/'))
    .replace(/[\\/]+/g, '/')
    .replace(/^[/]?/, '/')
    .replace(/[/]$/, '')
  return result
}


/**
 * Resolve local data path
 *
 * @param basePath
 * @param p
 */
export function resolveLocalPath(
  basePath: string,
  ...pathPieces: string[]
): string {
  const result = path.resolve(basePath, ...pathPieces)
  return result
}


/**
 * Calculate filepath cross platform and repository
 *
 * @param workspace
 * @param filepath
 */
export function resolveUniversalPath(
  workspace: string,
  filepath: string,
): string {
  const result = path
    .normalize(path.relative(workspace, path.resolve(workspace, filepath)))
    .replace(/[\\/]+/g, '/')
    .replace(/[/]$/, '')
  return result
}
