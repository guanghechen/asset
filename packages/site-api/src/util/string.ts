/**
 * Stringify data
 * @param data
 */
export function stringify(data: unknown): string {
  return JSON.stringify(data, null, 2)
}
