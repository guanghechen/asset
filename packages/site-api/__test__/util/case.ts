import path from 'path'
import {
  SiteConfig,
  createDefaultSubSiteConfig,
  loadSiteConfig,
  resolveSubSiteConfig,
  resolveSubSiteSourceItem,
} from '../../src'

const caseRootDir = path.resolve(__dirname, '../cases')
const resolveCasePath = (...relativePieces: string[]): string => (
  path.resolve(caseRootDir, ...relativePieces))


export interface CaseItem {
  /**
   * Case title
   */
  title: string
  /**
   * Root dir of the case
   */
  workspace: string
  /**
   * Site config provider
   */
  getSiteConfig: () => SiteConfig
}


export const cases: CaseItem[] = [
  // load full config
  ((): CaseItem => {
    const title = 'full-config'
    const workspace = resolveCasePath(title)
    const configFilepath = path.resolve(workspace, '_config.yml')

    const getSiteConfig = (): SiteConfig => {
      return loadSiteConfig(
        workspace,
        configFilepath,
        undefined,
        undefined,
        {
          barusu: function (rawConfig, sitePathConfig, defaultConfig) {
            return resolveSubSiteConfig(
              rawConfig,
              ['image'],
              resolveSubSiteSourceItem,
              sitePathConfig,
              {
                ...createDefaultSubSiteConfig('barusu'),
                source: {
                  image: {
                    sourceRoot: 'asset/image',
                    dataRoot: 'asset/image',
                    pattern: ['**/*{.png,jpg}'],
                  }
                },
              },
            )
          }
        }
      )
    }

    return { title, workspace, getSiteConfig }
  })()
]
