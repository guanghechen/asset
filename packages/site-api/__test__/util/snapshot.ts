import path from 'path'


const WORKSPACE_ROOT_DIR = path.resolve(__dirname, '../..')


/**
 * Remove sensitive data
 *
 * @param data
 */
export function desensitize<D extends unknown = unknown>(data: D): D {
  // No replaceAll, so use .split().join as a alternative
  const str = JSON.stringify(data)
    .split(WORKSPACE_ROOT_DIR).join('<PACKAGE_ROOT_DIR>')
  return JSON.parse(str)
}
