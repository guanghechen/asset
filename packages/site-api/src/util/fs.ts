import fs from 'fs-extra'
import path from 'path'

/**
 * Ensure the dirpath exists
 * @param dirpath
 */
export function ensurePathExists(dirpath: string): void {
  if (!fs.existsSync(dirpath)) {
    fs.mkdirsSync(dirpath)
  }
}

/**
 * Output into file
 *
 * @param filepath
 * @param data
 */
export async function writeFile(filepath: string, data: Buffer): Promise<void> {
  ensurePathExists(path.dirname(filepath))
  await fs.writeFile(filepath, data)
}

/**
 * Output into file (synchronous)
 *
 * @param filepath
 * @param data
 */
export function writeFileSync(filepath: string, data: Buffer): void {
  ensurePathExists(path.dirname(filepath))
  fs.writeFileSync(filepath, data)
}

/**
 * Output into json file
 *
 * @param filepath
 * @param data
 */
export async function writeJSON<T extends unknown = unknown>(
  filepath: string,
  data: T,
): Promise<T> {
  ensurePathExists(path.dirname(filepath))
  await fs.writeJSON(filepath, data)
  return data
}

/**
 * Output into json file (synchronous)
 *
 * @param filepath
 * @param data
 */
export function writeJSONSync<T extends unknown = unknown>(
  filepath: string,
  data: T,
): T {
  ensurePathExists(path.dirname(filepath))
  fs.writeJSONSync(filepath, data)
  return data
}
