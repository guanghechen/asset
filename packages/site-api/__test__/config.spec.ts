import path from 'path'
import type { SubSiteConfig } from '../src'
import { createDefaultSubSiteConfig, resolveSubSiteConfig, resolveSubSiteSourceItem } from '../src'
import { cases } from './util/case'
import { desensitize } from './util/snapshot'

describe('resolveSubSiteConfig', function () {
  const resolveConfig = (
    workspace: string,
    rawSubSiteConfig: Partial<SubSiteConfig>,
  ): SubSiteConfig => {
    const sourceItemTypes = ['image']
    const resolveSourceItem = resolveSubSiteSourceItem
    const sitePathConfig = { routeRoot: '/', urlRoot: '/', workspace }
    const defaultConfig: SubSiteConfig = {
      ...createDefaultSubSiteConfig('demo'),
      source: {
        image: {
          sourceRoot: 'asset/image',
          dataRoot: 'asset/image',
          pattern: ['**/*{.png,jpg}'],
        },
      },
    }

    const subSiteConfig = resolveSubSiteConfig(
      rawSubSiteConfig,
      sourceItemTypes,
      resolveSourceItem,
      sitePathConfig,
      defaultConfig,
    )
    return subSiteConfig
  }

  test('empty rawConfig', function () {
    expect(desensitize(resolveConfig(path.resolve(__dirname, 'case'), {}))).toMatchSnapshot()
  })

  test('full rawConfig', function () {
    expect(
      desensitize(
        resolveConfig(path.resolve(__dirname, 'case'), {
          routeRoot: '/waw',
          sourceRoot: 'emm/waw/source',
          dataRoot: 'waw/emm/data',
          assetDataMapFilepath: 'asset.json',
          tagDataMapFilepath: 'tag.json',
          categoryDataMapFilepath: 'category.json',
        }),
      ),
    ).toMatchSnapshot()
  })
})

describe('loadSiteConfig', function () {
  for (const kase of cases) {
    const { title, getSiteConfig } = kase
    // eslint-disable-next-line jest/valid-title
    test(title, function () {
      const siteConfig = getSiteConfig()
      expect(desensitize(siteConfig)).toMatchSnapshot()

      // loadSiteConfig should be a pure function
      expect(getSiteConfig()).toEqual(siteConfig)
    })
  }
})
