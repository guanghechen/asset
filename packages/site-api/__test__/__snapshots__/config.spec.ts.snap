// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadSiteConfig full-config 1`] = `
Object {
  "author": "Demo author",
  "deploy": Object {
    "branch": "master",
    "email": "",
    "name": "",
    "repository": "",
    "type": "git",
  },
  "description": "Demo description",
  "sites": Object {
    "barusu": Object {
      "assetDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data/asset.map.json",
      "categoryDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data/category.map.json",
      "dataRoot": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data",
      "source": Object {
        "image": Object {
          "dataRoot": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data/image",
          "pattern": Array [
            "**/*{.png,jpg}",
          ],
          "sourceRoot": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/source",
        },
      },
      "sourceRoot": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/source",
      "tagDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config/barusu/data/tag.map.json",
      "urlRoot": "/barusu",
    },
  },
  "title": "Demo title",
  "urlRoot": "/",
  "workspace": "<PACKAGE_ROOT_DIR>/__test__/cases/full-config",
}
`;

exports[`resolveSubSiteConfig empty rawConfig 1`] = `
Object {
  "assetDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/case/demo/data/asset.map.json",
  "categoryDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/case/demo/data/category.map.json",
  "dataRoot": "<PACKAGE_ROOT_DIR>/__test__/case/demo/data",
  "source": Object {
    "image": Object {
      "dataRoot": "<PACKAGE_ROOT_DIR>/__test__/case/demo/data/asset/image",
      "pattern": Array [
        "**/*{.png,jpg}",
      ],
      "sourceRoot": "<PACKAGE_ROOT_DIR>/__test__/case/demo/source/asset/image",
    },
  },
  "sourceRoot": "<PACKAGE_ROOT_DIR>/__test__/case/demo/source",
  "tagDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/case/demo/data/tag.map.json",
  "urlRoot": "/demo",
}
`;

exports[`resolveSubSiteConfig full rawConfig 1`] = `
Object {
  "assetDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/case/waw/emm/data/asset.json",
  "categoryDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/case/waw/emm/data/category.json",
  "dataRoot": "<PACKAGE_ROOT_DIR>/__test__/case/waw/emm/data",
  "source": Object {
    "image": Object {
      "dataRoot": "<PACKAGE_ROOT_DIR>/__test__/case/waw/emm/data/asset/image",
      "pattern": Array [
        "**/*{.png,jpg}",
      ],
      "sourceRoot": "<PACKAGE_ROOT_DIR>/__test__/case/emm/waw/source/asset/image",
    },
  },
  "sourceRoot": "<PACKAGE_ROOT_DIR>/__test__/case/emm/waw/source",
  "tagDataMapFilepath": "<PACKAGE_ROOT_DIR>/__test__/case/waw/emm/data/tag.json",
  "urlRoot": "/waw",
}
`;
